strings/base_types.h

```diff
@@ -170,16 +170,27 @@ WINRT_EXPORT namespace winrt
         }
     };
 
-    inline bool operator==(guid const& left, guid const& right) noexcept
+    constexpr bool operator==(guid const& left, guid const& right) noexcept
     {
-        return !memcmp(&left, &right, sizeof(left));
+        return
+            (left.Data1 == right.Data1) &&
+            (left.Data2 == right.Data2) &&
+            (left.Data3 == right.Data3) &&
+            (left.Data4[0] == right.Data4[0]) &&
+            (left.Data4[1] == right.Data4[1]) &&
+            (left.Data4[2] == right.Data4[2]) &&
+            (left.Data4[3] == right.Data4[3]) &&
+            (left.Data4[4] == right.Data4[4]) &&
+            (left.Data4[5] == right.Data4[5]) &&
+            (left.Data4[6] == right.Data4[6]) &&
+            (left.Data4[7] == right.Data4[7]);
     }
```

Comment:
Hmm, I'm worried this may have an impact on code size. `memcmp` was carefully picked to reduce code size as it is used repeatedly in every QueryInterface implementation. We'd have to build something big (like WinUI) with LTCG to see whether it remains an issue.

Reply:
@DefaultRyan may have some more recent experience with this. My data is pretty old.

Reply:
Ah, ok; I wasn't aware of those considerations. On the one hand, I'm curious and hopeful that the compiler can be smart about this esp. since it's `constexpr`. On the other hand, the motivation is a it weak; I only changed this to make the test easier to write. So it's probably best to revert this part and hack the test together instead.

Reply:
I'd recommend `if (std::is_constant_evaluated())` for this, but then that's C++20 only.

Reply:
Just tested if the compiler was smart about it, it isn't: compiling with `/FA` gives this

![image](https://user-images.githubusercontent.com/6440374/127718202-9c581fd1-fd73-447f-8483-a9a8a2193fe1.png)

Reply:
Thanks for looking into it! I'll revert this part of the change for now.